tools/file_packager.py

```diff
@@ -191,17 +198,14 @@
 if not has_preloaded or jsoutput == None:
   assert not separate_metadata, 'cannot separate-metadata without both --preloaded files and a specified --js-output'
 
-ret = '''
-var Module;
-'''
-if not no_closure:
-  ret += '''
-if (typeof Module === 'undefined') Module = eval('(function() { try { return Module || {} } catch(e) { return {} } })()');
-'''
```

Comment:
Thanks, this looks better and clearer. I still there's something wrong here though. There are 3 cases:

1. We run with closure. We must be called from emcc, and so can do `Module = [pattern]`, as it's going to be replaced anyhow.
2. We run without closure, called from emcc.py. This will do `Module = Module || {}` (or export name), which is also ok.
3. We run without closure, **not** called from emcc.py. It is not valid in this case to do `Module = ` as there is no `var Module;` that corresponds to us: we are in another script tag entirely from emcc output.

This is a little tricky, I'm afraid. And it also has to do with

https://github.com/kripken/emscripten/issues/5900#issuecomment-349453563

which complicates matters further.

I think (not 100% sure) we can fix this as follows:

 * The file packager needs to get a param "am I being called from emcc.py". Then it can differentiate the 2nd and 3rd cases above. In the third, it can emit `var Module`.
 * We can solve the issue from #5900 and simplify the code here by making the following change: the logic that "gets" the `Module` object, that the file packager is duplicating really, can be done in MODULARIZE mode **before** pre-js code (which would include file packager code). So in this mode we wouldn't do it in shell.js, or in the file packager (however, when run outside of emcc.py, the file packager would need something). But maybe we also need to tell the file packager if we are in modularize mode (and not just emcc mode).

I realize this is getting convoluted, sorry about that. If you want, we can land this into a side branch and I can investigate the two bullet points above.

Reply:
As for 3 I'm pretty sure there is either `var Module` somewhere or it is not covered by any test case yet. Otherwise I'd not remove `var` from there. These duplicated `var` all over the place are quire difficult to trace. I'm still not 100% aware of what is calling where and under what circumstances.

Properly solving #5900 is a better route I think. However, if this PR doesn't make #5900 worse I'd rather land this and work on #5900 specifically in a separate PR.

Reply:
I see, this PR will throw away the code that was "fixing" #5900.

Reply:
Hmm, I'm tired and may have overcomplicated things. The second bullet point can probably be just a little reordering of how MODULARIZE works - it's already different in that pre-js is not before all the code (since now the code is inside a function). Might be a simple and natural fix there, I hope.